---
title: "Diff_exp_linear_mixed_model_limma"
output: html_document
date: "2025-06-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:



```{r}
phenotype = read.csv(file="PI study gene expression IDs from Jiaying.csv",header = T,stringsAsFactors = F)
confounder = read.csv(file="Covariates_info.csv",header = T,stringsAsFactors = F)
head(phenotype)

phenotype = phenotype[,c(1:6)]
#phenotype_1 = phenotype[,c(1:6)]
#phenotype_2 = phenotype[,c("Box.nr.","ID","Visit","Novogene.ID","Meal.time","Meal_type")]

phenotype$Diet_name[phenotype$Meal_type=="A"] = "Corn_oil"
phenotype$Diet_name[phenotype$Meal_type=="B"] = "Butter"
phenotype$Diet_name[phenotype$Meal_type=="C"] = "Flaxseed_oil"
phenotype$Diet_name[phenotype$Meal_type=="D"] = "Coconut_oil"

head(phenotype)

table(phenotype$Diet_name,phenotype$Meal_type)
#table(phenotype$Meal_type,phenotype$Diet_name)

class(confounder$Gender)
confounder$Gender = as.factor(confounder$Gender)


path_to_file="/Users/siddhaduio.no/Desktop/PhD_Project_related/PI_dietary_intervention_study_gene_Exp_data"


files_ordered = paste(path_to_file,phenotype$Novogene.ID,sep="/")
files_ordered = paste(files_ordered,"quant.sf",sep="/")

class(files_ordered)
files_ordered = as.data.frame(files_ordered)

library(stringr)
files_ordered[paste("Col",1:7,sep="")] <- str_split_fixed(files_ordered$files_ordered, '/', 8)

files_ordered=files_ordered[,c("files_ordered","Col7")]

colnames(phenotype)[4] = "Novogene_ID"

library(dplyr)
phenotype_ordered = left_join(files_ordered,phenotype,by=c("Col7"="Novogene_ID"))
phenotype_confounder = left_join(phenotype_ordered,confounder,by=c("ID"="Study.ID"))

phenotype_confounder$Age_c = scale(phenotype_confounder$Age,center = TRUE)
phenotype_confounder$BMI_c = scale(phenotype_confounder$BMI.from.Screening,center = TRUE)
phenotype_confounder[paste("Visit_split",1:2,sep="")] <- str_split_fixed(phenotype_confounder$Visit, '-', 2)
colnames(phenotype_confounder)[14]="Visit"
colnames(phenotype_confounder)[15]="Time"
phenotype_confounder$Combined_var = paste(phenotype_confounder$Time,phenotype_confounder$Diet_name,sep="_")

```

Now reading in the gene exp data

```{r}
library(tibble)

res=read.table(file="GRChg38_gene_and_transcript_stable_ID_version.txt",header = TRUE,sep="\t",stringsAsFactors = FALSE)
res=res[,c(1,4)]
colnames(res)=c("GENEID","TXNAME")
res=res[,c(2,1)]
res_tibble=as_tibble(res)


library(stringr)
library(DESeq2)
library(tximport)

`%ni%` = Negate(`%in%`)

phenotype_no_AJ_5 = subset(phenotype_confounder,phenotype_confounder$Col7 %ni% c("AJ_5"))

butter = subset(phenotype_no_AJ_5,phenotype_no_AJ_5$Combined_var=="T0_Butter" | phenotype_no_AJ_5$Combined_var=="T4_Butter")

coconut_oil = subset(phenotype_no_AJ_5,phenotype_no_AJ_5$Combined_var=="T0_Coconut_oil" | phenotype_no_AJ_5$Combined_var=="T4_Coconut_oil")

corn_oil = subset(phenotype_no_AJ_5,phenotype_no_AJ_5$Combined_var=="T0_Corn_oil" | phenotype_no_AJ_5$Combined_var=="T4_Corn_oil")

flaxseed_oil = subset(phenotype_no_AJ_5,phenotype_no_AJ_5$Combined_var=="T0_Flaxseed_oil" | phenotype_no_AJ_5$Combined_var=="T4_Flaxseed_oil")



rnaseq_files = phenotype_no_AJ_5$files_ordered

txi = tximport(rnaseq_files,type = "salmon",tx2gene = res_tibble,countsFromAbundance = "no")
class(txi$counts) ## a total of 43 samples
head(txi$counts)
colnames(txi$counts)= phenotype_no_AJ_5$Col7
head(txi$counts)

```

#Now use glmmSeq for RNA -seq mixed effect model

edgeR to get dispersion
```{r}
## model fit for 
#dds <- DESeqDataSetFromTximport(txi = txi, colData = phenotype_no_AJ_5, design = ~ 1)
#dds <- DESeq(dds)
#dispersions <- setNames(dispersions(dds), rownames(txi$counts))

#head(dispersions)
```
Size Factors

```{r}
#sizeFactors <- estimateSizeFactorsForMatrix(txi$counts)  # normalise to mean = 1

#head(sizeFactors)
```


```{r}

## Nothing here

#dds = dds[1:1000,]
```


Limma Analysis
Limma has a built-in approach for analyzing repeated measures data using duplicateCorrelation(). The model can handle a single random effect, and forces the magnitude of the random effect to be the same across all genes.


```{r}






```

First process wit edgeR

```{r}
library(edgeR)
# filter genes by number of counts
isexpr <- rowSums(cpm(txi$counts) > 0.5) >= 5

# Standard usage of limma/voom
dge <- DGEList(txi$counts[isexpr, ])
dge <- calcNormFactors(dge)

# make this vignette faster by analyzing a subset of genes
#dge <- dge[1:1000, ]
```
Limma analysis
Limma has a built-in approach for analyzing repeated measures data using duplicateCorrelation(). The model can handle a single random effect, and forces the magnitude of the random effect to be the same across all genes.

```{r}
# apply duplicateCorrelation is two rounds
design <- model.matrix(~ Combined_var + BMI_c + Age_c + Gender, phenotype_no_AJ_5)
phenotype_no_AJ_5$ID = as.character(phenotype_no_AJ_5$ID)
phenotype_no_AJ_5$Combined_var = as.factor(phenotype_no_AJ_5$Combined_var)
vobj_tmp <- voom(dge, design, plot = FALSE)
dupcor <- duplicateCorrelation(vobj_tmp, design, block = phenotype_no_AJ_5$ID)

# run voom considering the duplicateCorrelation results
# in order to compute more accurate precision weights
# Otherwise, use the results from the first voom run
vobj <- voom(dge, design, plot = FALSE, block = phenotype_no_AJ_5$ID, correlation = dupcor$consensus)

# Estimate linear mixed model with a single variance component
# Fit the model for each gene,
dupcor <- duplicateCorrelation(vobj, design, block = phenotype_no_AJ_5$ID)

# But this step uses only the genome-wide average for the random effect
fitDupCor <- lmFit(vobj, design, block = phenotype_no_AJ_5$Individual, correlation = dupcor$consensus)

# Fit Empirical Bayes for moderated t-statistics
fitDupCor <- eBayes(fitDupCor)
```

Dream Analysis
The dream method replaces 4 core functions of limma with a linear mixed model.

voomWithDreamWeights() replaces voom() to estimate precision weights
dream() replaces lmFit() to estimate regression coefficients.
variancePartition::eBayes() replaces limma::eBayes() to apply empircal Bayes shrinkage on linear mixed models.
variancePartition::topTable() replaces limma::topTable() to give seamless access to results from dream().
For models with only fixed effects, variancePartition::eBayes(), and variancePartition::topTable() work seamlessly and give results equivalent to the limma functions with the same name. From the user perspective, the dream() workflow is the same as limma since the statistical differences are handled behind the scenes.

```{r}
# Specify parallel processing parameters
# this is used implicitly by dream() to run in parallel
rownames(phenotype_no_AJ_5) = phenotype_no_AJ_5$Col7
library(BiocParallel)
library(variancePartition)
param <- SnowParam(4, "SOCK", progressbar = TRUE)

# The variable to be tested must be a fixed effect
form <- ~ Combined_var + BMI_c + Age_c + Gender +  (1 | ID)

# estimate weights using linear mixed model of dream
vobjDream <- voomWithDreamWeights(dge, form, phenotype_no_AJ_5, BPPARAM = param)

# Fit the dream model on each gene
# For the hypothesis testing, by default,
# dream() uses the KR method for <= 20 samples,
# otherwise it uses the Satterthwaite approximation
fitmm <- dream(vobjDream, form, phenotype_no_AJ_5)
fitmm <- eBayes(fitmm)
```
# CHeck design matrix

```{r}
head(fitmm$design, 3)
```
```{r}
 #Get results of hypothesis test on coefficients of interest
topTable(fitmm, coef = "Combined_varT4_Corn_oil", number = 3)
```
Using contrasts to compare coefficients
We can also perform a hypothesis test of the difference between two or more coefficients by using a contrast matrix. The contrasts are evaluated at the time of the model fit and the results can be extracted with topTable(). This behaves like makeContrasts() and contrasts.fit() in limma.

Multiple contrasts can be evaluated at the same time, in order to save computation time. Make sure to inspect your contrast matrix to confirm it is testing what you intend.

```{r}
phenotype_no_AJ_5$Combined_var = as.factor(phenotype_no_AJ_5$Combined_var)
form <- ~ 0 + Combined_var + Gender +  Age_c + BMI_c + (1 | ID)

L <- makeContrastsDream(form, phenotype_no_AJ_5,
  contrasts = c(
    compare_butter_T4_vs_T0 = "Combined_varT4_Butter - Combined_varT0_Butter",
    compare_flaxseed_T4_vs_T0 = "Combined_varT4_Flaxseed_oil - Combined_varT0_Flaxseed_oil",
    compare_coconut_T4_vs_T0 =  "Combined_varT4_Coconut_oil - Combined_varT0_Coconut_oil",
    compare_cornoil_T4_vs_T0 = "Combined_varT4_Corn_oil - Combined_varT0_Corn_oil"
  )
)

# Visualize contrast matrix
plotContrasts(L)
```
fit model with contrasts

```{r}
# fit dream model with contrasts
fit_model <- dream(vobjDream, form, phenotype_no_AJ_5, L)
fit_model <- eBayes(fit_model)

# get names of available coefficients and contrasts for testing
colnames(fit_model)
```
Joint hypothesis test of multiple coefficients
Joint hypothesis testing of multiple coefficients at the same time can be performed by using an F-test. Just like in limma, the results can be extracted using topTable()


```{r}
cornoil_deg_res=topTable(fit_model, coef = c("compare_cornoil_T4_vs_T0"), number = 16542)
butter_deg_res=topTable(fit_model, coef = c("compare_butter_T4_vs_T0"), number = 16542)
flaxseed_deg_res=topTable(fit_model, coef = c("compare_flaxseed_T4_vs_T0"), number = 16542)
coconut_deg_res=topTable(fit_model, coef = c("compare_coconut_T4_vs_T0"), number = 16542)

```
variancePartition plot
Dream and variancePartition share the same underlying linear mixed model framework. A variancePartition analysis can indicate important variables that should be included as fixed or random effects in the dream analysis.

```{r}
form <- ~ (1 | ID) + (1 | Combined_var)
vp <- fitExtractVarPartModel(vobj, form, phenotype_no_AJ_5)

plotVarPart(sortCols(vp))
```

Comparing p values
```{r}
# Compare p-values and make plot
p1 <- topTable(fitDupCor, coef = "Combined_varT0_Coconut_oil", number = Inf, sort.by = "none")$P.Value
p2 <- topTable(fitmm, number = Inf, sort.by = "none")$P.Value

plotCompareP(p1, p2, vp$ID, dupcor$consensus)
```

Write to file the tables

```{r}
write.table(cornoil_deg_res,file="Cornoil_T4_vs_T0_linear_mixed_model_adjusted_age_gender_BMI_using_dream_and_limma.txt",col.names = T,row.names = T,sep="\t",quote = F)
write.table(butter_deg_res,file="Butter_T4_vs_T0_linear_mixed_model_adjusted_age_gender_BMI_using_dream_and_limma.txt",col.names = T,row.names = T,sep="\t",quote = F)
write.table(flaxseed_deg_res,file="Flaxseed_T4_vs_T0_linear_mixed_model_adjusted_age_gender_BMI_using_dream_and_limma.txt",col.names = T,row.names = T,sep="\t",quote = F)
write.table(coconut_deg_res,file="Coconut_T4_vs_T0_linear_mixed_model_adjusted_age_gender_BMI_using_dream_and_limma.txt",col.names = T,row.names = T,sep="\t",quote = F)

```

```{r}
pdf(file="Violin_plot_of_variance_fraction_for_each_gene_and_each_variable.pdf")
plotVarPart(sortCols(vp))
dev.off()
```

